CARDS-1417: Print-friendly styling of forms - refactoring with a reusable, self-contained Print button

veronikaslc opened the pull request.

Commit:
CARDS-1417: Print-friendly styling of forms

Commit:
[misc] Fixed missing unique key prop console error via not generating empty Lists

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/PrintPreview.jsx:
```diff
@@ -145,14 +157,16 @@ const PrintPreview = forwardRef((props, ref) => {
       <DialogActions>
         <Button variant="contained" onClick={onClose}>Close</Button>
-        <Button variant="contained" color="primary" onClick={() => {window.print()}} disabled={!!!content}>Print</Button>
+        <Button variant="contained" onClick={handlePrint} disabled={!!!content}>Print</Button>
```
`color="primary"` should not have been removed
```suggestion
        <Button variant="contained" color="primary" onClick={handlePrint} disabled={!!!content}>Print</Button>
```

marta- reviewed (changes requested):
Splitting a long form into pages appears broken:
![image](https://user-images.githubusercontent.com/651980/151269640-59715125-2a0f-4f98-ac2f-1e333553e974.png)

If also generates a gray block at the end of the content if there's any space left on the last page.

sashaandjic reviewed (approved):
**Question:** Can we have this functionality [here](https://user-images.githubusercontent.com/23243244/151276026-94ed1714-30ab-406a-a049-c49d797613f2.png) as well?
Or at least option to email the results back to the patient? Both?

sashaandjic commented:
Patient chart looks terrible in this PR. Hope this is completely out of scope.
![screencapture-localhost-8080-content-html-Subjects-beca1d9c-d086-4a19-b1a9-62ee1949e2c1-e7b77324-e3b5-44d3-a833-8342c4e8c2e9-2022-01-26-20_46_15](https://user-images.githubusercontent.com/23243244/151277018-baa79c12-33ae-41e5-a351-0cfc0ca33e2d.png)

sashaandjic commented:
Should `PRINT` button be dark blue?
![image](https://user-images.githubusercontent.com/23243244/151277842-c7ee2074-1948-4564-a5c8-10dbfcf8ad54.png)

marta- commented:
> Patient chart looks terrible in this PR. Hope this is unrelated and completely out of scope.

It is unrelated. It is caused by the bulky computed interpretations that we have in the surveys.

marta- commented:
> **Question:** Can we have this functionality here as well? Or at least option to email the results back to the patient? Both?

Please make a separate task. The implementation would have to be very different from what we have here, and it's low priority since it wasn't requested in the specs.

Commit:
CARDS-1417: Print-friendly styling of forms - refactoring with a reusable, self-contained Print button
Added back the button promary color property

sashaandjic commented:
The dark blue button looks great.
Splitting a long form into pages is still broken
[CARDS-1417 _ 1_ Patient Health Questionnaire-9 _ DATA-PRO.pdf](https://github.com/data-team-uhn/cards/files/7947923/CARDS-1417._.1_.Patient.Health.Questionnaire-9._.DATA-PRO.pdf)
.

sashaandjic commented:
Can we make the top and the bottom line red?
![image](https://user-images.githubusercontent.com/23243244/151300312-0baacd63-cd98-4b8d-9872-e58fb968282d.png)

sashaandjic commented:
It would be nice if we can remove this information from view mode and consequently from the print preview. 
![image](https://user-images.githubusercontent.com/23243244/151301593-0a50cc0c-7e5c-427c-9c2a-37bfe7d1abd5.png)

marta- commented:
> It would be nice if we can remove this information from view mode and consequently from the print preview.

This is what the question looks like ("Here is some information...") and that is the patient's answer which needs to be printed out, We can't change the specs of the Smoking cessation questionnaire.

marta- commented:
> Can we make the top and the bottom line red?

Why red? That's usually reserved for highlighting something dangerous. Also, the top red line would get printed out. Did you want the parts that don't get printed (the buttons zone at the bottom) out styled differently instead, e.g. with a gray background?

sashaandjic commented:
> Why red? That's usually reserved for highlighting something dangerous. Also, the top red line would get printed out. Did you want the parts that don't get printed (the buttons zone at the bottom) out styled differently instead, e.g. with a gray background?

It looks a bit to much B&W. How about dark blue?

sashaandjic commented:
> This is what the question looks like ("Here is some information...") and that is the patient's answer which needs to be printed out, We can't change the specs of the Smoking cessation questionnaire.

How about moving "Here is some information about each program. You can only choose one program." into an info box, and changing question label to: "Click on the program you want to register in."?

marta- commented:
> > This is what the question looks like ("Here is some information...") and that is the patient's answer which needs to be printed out, We can't change the specs of the Smoking cessation questionnaire.
> 
> How about moving "Here is some information about each program. You can only choose one program." into an info box, and changing question label to: "Select the program you want to register in."?

Please make a separate [IDEA] task. This is unrelated to printing implementation.

marta- commented:
> > Why red? That's usually reserved for highlighting something dangerous. Also, the top red line would get printed out. Did you want the parts that don't get printed (the buttons zone at the bottom) out styled differently instead, e.g. with a gray background?
> 
> It looks a bit too much B&W and kind of old. How about dark blue?

Noted. We are currently working on fixing the page break issue, which might involve changing what parts of the dialog get printed (whether header is included). After this is established, we can move along to styling the parts that do not get printed in a distinctive way. So let's revisit styling after the bug is fixed.

marta- committed:
CARDS-1417: Print-friendly styling of forms - refactoring with a reusable, self-contained Print button
Bugfix: page breaks sometimes split a line of text vertically between two pages
(this bug was present already in the original implementation of 1417).
To fix, instead of printing the Preview dialog, whose styling is screen-optimized and caused the vertical split,
added a separate print target (<Card>) which replicates the print preview without the faulty layout and is
not visible in the frontend on `media @screen`.

marta- committed:
CARDS-1417: Print-friendly styling of forms - refactoring with a reusable, self-contained Print button
Added an option to the print button to skip through the print preview dialog and go straight to the browser-native print function
Note: with the current code, the dropdown menu remains open after printing is completed.

marta- committed:
CARDS-1417: Print-friendly styling of forms - refactoring with a reusable, self-contained Print button
[Code style] Improved indentation

marta- committed:
CARDS-1417: Print-friendly styling of forms - refactoring with a reusable, self-contained Print button
Set the default @page margin, smaller padding for the printed Card

marta- commented:
Notable changes:
* rebased on `CARDS-1599_clean` to include the latest markdown serialization changes
* `react-to-print` target is now  `Card` only visible in `@media print` instead of the print preview dialog - this fixes the page breaks mid-line
* set default margins to the `@page` for nicer-looking prints with default settings
* added the option to `disablePreview` to not show the preview dialog and skip directly to the browser's print preview. It doesn't work perfectly for now: there's a js console warning when this option is on, and the form's dropdown menu doesn't close when we're done printing, so the preview is not disabled by default. But I left the code in for perfecting at a later time (not a priority for this PR).

marta- commented:
Demoting to Draft only until #877  is merged.

sashaandjic reviewed (approved):
tested
Question: How can I manipulate the disablePreview option?
[CARDS-1417.pdf](https://github.com/data-team-uhn/cards/files/7965905/CARDS-1417.pdf)

marta- commented:
> tested Question: How can I manipulate the disablePreview option?

It's an option to the React component that renders the button. It can't be manipulated without changing the code.
